### ISSUE TITLE
autodiscovery

### ISSUE THREAD

**Issue by kplimack** [metal3-io/baremetal-operator#1124]
It would be a big time saver if `BareMetalHost` objects did not require the `BootMACAddress`, but instead were able to rely on the data collected during inspection.  I would love to be able to intake large quantities of hardware and not need to first collect all of their MACs.  When the onboard MAC isnt used, its not easy to ask a hardware vendor to provide it.

**Comment by kplimack** [comment]
seems like this discussion has come up before, https://github.com/metal3-io/baremetal-operator/pull/284

When I have hardware to intake, I know all the BMC MAC Addrs and serve static-DHCP via ironic-dnsmasq and a dhcp-relay, so I can see all of the nodes booting and identify them.  Ive also had success scraping the MAC addresses using redfish, though sometimes it is returned as 00:00:00:00:00.  
Perhaps we can try detecting the mac addresses and performing discovery during the `registration` phase, and fail registration if we get all 0's, but move on to inspection if good data is recieved

**Comment by kplimack** [comment]
also https://github.com/metal3-io/baremetal-operator/issues/979

**Comment by zaneb** [comment]
Not every driver requires the Boot MAC, but the ones that do (all PXE drivers aiui) require it in order to boot the IPA image that does inspection. So there's no way to get it from inspection.

**Comment by dtantsur** [comment]
I think we're conflating two things here:

1) Not requiring a boot MAC address. Ironic supports that for Redfish: it automatically does OOB inspection before running a full one. We may need to test it and remove the requirement, if needed. Nothing like that for IPMI.

2) Auto-discovery, i.e. the process of automatically enrolling nodes that PXE boot from our DHCP server. This is supported by ironic-inspector, but is optional and not enabled for Metal3. The reason is that BMO will not be aware of the newly created nodes. There is also a caveat: we cannot discovery BMC credentials, they'll need to be populated afterwards.

Which of the options are we talking about?

**Comment by kplimack** [comment]
@dtantsur am I misunderstanding this code?
https://github.com/metal3-io/baremetal-operator/blob/main/pkg/hardwareutils/bmc/redfish.go#L54-L58
It seems like redfish does require the boot mac address

**Comment by dtantsur** [comment]
Yes, because the work on the BMO side has not been done yet.

> We may need to test it and remove the requirement

**Comment by dtantsur** [comment]
/triage needs-information

So, what is it about, Redfish or generic auto-discovery?

**Comment by furkatgofurov7** [comment]
> So, what is it about, Redfish or generic auto-discovery?

cc @kplimack

**Comment by ReggieCarey** [comment]
Canonicals MAAS tool provisions bare metal and discovers them when they're booted onto the IPMI network - so it can be done.  MAAS allows manual adding of machines and requests bootMACAddress and IPMI endpoint & credentials if you're manually adding records.  For auto discovered machines, we get a full inspection. I suspect this is more of a limitation of IRONIC than anything else. I find that ipmitool used by Ironic has issues with machines that MAAS has no issues with (cipher selection issues primarily).

**Comment by dtantsur** [comment]
@ReggieCarey 

> Canonicals MAAS tool provisions bare metal and discovers them when they're booted onto the IPMI network - so it can be done.

Yep, Ironic (more precisely, Inspector) can do it (but not in the way Metal3 configures it).

What we do not do that MAAS does is auto-generating IPMI credentials. It has been researched, it has unpleasant security and stability implications. You can achieve that with some sort of a custom ramdisk if you really want to.

What you can do in Inspector (but not baremetal-operator) is to write rules that assign credentials, e.g. based on vendor.

> I suspect this is more of a limitation of IRONIC than anything else.

Nope, rather of baremetal-operator really (modulo the credentials question). BMO expects to own all Ironic nodes and will not recognize newly discovered nodes. You'll need to enroll BareMetalHosts explicitly, which defeats the purpose.

> I find that ipmitool used by Ironic has issues with machines that MAAS has no issues with (cipher selection issues primarily).

It depends on the version of ipmitool, not Ironic. The difference you observe is probably because our images are based on CentOS Stream, while MAAS is Ubuntu.


So, TL;DR some of this is doable, it's a bit of work on the BMO side, none of us have time for it in the near future.

**Comment by furkatgofurov7** [comment]
cc @kplimack can we close this issue or is there any comment for the above statements from @dtantsur ?

**Comment by dtantsur** [comment]
We've had a discussion about discovery in Metal3 on the Ironic IRC, so here is my take on how this feature may have a future:

We can make auto-discovery opt-in in both BMO and ironic-image (it's already an option in the latter). When it is on, Ironic will enroll any new hosts that boot into its PXE environment. Then we add a Metal3-specific inspection plugin into ironic-image that creates corresponding BareMetalHost records [*] with no credentials.

The credentials question is explicitly left unsolved. The recommended way would be to write a downstream controller that listens for hosts in the "unmanaged" state and updates them e.g. from a CMDB. The not quite recommended way would be a custom IPA image that initializes credentials. Up to the consumer, Metal3 won't dictate the right approach.

In the end, I'm not going to design and implement such a feature, but I'm open to reviewing a design proposal that takes this and all the previous arguments into account.

[*] Accessing Kubernetes from Ironic? Yes, some folks may not opt into that for security reasons. But otherwise BMO needs to duplicate options for the default driver name, for instance. And we need a periodic task in BMO for polling Ironic rather then using the normal reconcile process.
